extensions-contrib/protobuf-extensions/src/main/java/io/druid/data/input/protobuf/ProtoBufInputRowParser.java

```diff
@@ -0,0 +1,138 @@
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.data.input.protobuf;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.annotation.JsonTypeName;
+import com.github.os72.protobuf.dynamic.DynamicSchema;
+import com.google.protobuf.ByteString;
+import com.google.protobuf.Descriptors;
+import com.google.protobuf.Descriptors.Descriptor;
+import com.google.protobuf.DynamicMessage;
+import com.google.protobuf.InvalidProtocolBufferException;
+import com.google.protobuf.util.JsonFormat;
+import io.druid.data.input.ByteBufferInputRowParser;
+import io.druid.data.input.InputRow;
+import io.druid.data.input.MapBasedInputRow;
+import io.druid.data.input.impl.ParseSpec;
+import io.druid.java.util.common.parsers.ParseException;
+import io.druid.java.util.common.parsers.Parser;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.nio.ByteBuffer;
+import java.util.Map;
+
+@JsonTypeName("protobuf")
+public class ProtoBufInputRowParser implements ByteBufferInputRowParser {
+  //private static final Logger log = new Logger(ProtoBufInputRowParser.class);
+
+  private final ParseSpec parseSpec;
+  private Parser<String, Object> parser;
+  private final String descriptorFilePath;
+  private final String protoMessageType;
+  private Descriptor descriptor;
+
+
+  @JsonCreator
+  public ProtoBufInputRowParser(
+      @JsonProperty("parseSpec") ParseSpec parseSpec,
+      @JsonProperty("descriptor") String descriptorFilePath,
+      @JsonProperty("protoMessageType") String protoMessageType
+  ) {
+    this.parseSpec = parseSpec;
+    this.descriptorFilePath = descriptorFilePath;
+    this.protoMessageType = protoMessageType;
+    this.parser = parseSpec.makeParser();
+    this.descriptor = getDescriptor(descriptorFilePath);
+  }
+
+  @Override
+  public ParseSpec getParseSpec() {
+    return parseSpec;
+  }
+
+  @Override
+  public ProtoBufInputRowParser withParseSpec(ParseSpec parseSpec) {
+    return new ProtoBufInputRowParser(parseSpec, descriptorFilePath, protoMessageType);
+  }
+
+  @Override
+  public InputRow parse(ByteBuffer input) {
+    String json;
+    try {
+      DynamicMessage message = DynamicMessage.parseFrom(descriptor, ByteString.copyFrom(input));
+      json = JsonFormat.printer().print(message);
+      // log.trace(json);
+    } catch (InvalidProtocolBufferException e) {
+      throw new ParseException(e, "Protobuf message could not be parsed");
+    }
+
+    Map<String, Object> record = parser.parse(json);
```

Comment:
I'm well aware of the performance issue.  What I looked at is the [JsonProvider](https://github.com/jayway/JsonPath/blob/master/json-path/src/main/java/com/jayway/jsonpath/spi/json/JsonProvider.java) and [MappingProvider](https://github.com/jayway/JsonPath/blob/master/json-path/src/main/java/com/jayway/jsonpath/spi/mapper/MappingProvider.java) SPI in the JSONPath module.  The benefit is we could provide the `flattenSpec` not only for Protobuf but potentially any format e.g. AvroProvider as I discussed at #3505.

Unfortunately the provider interface requires [String](https://github.com/jayway/JsonPath/blob/master/json-path/src/main/java/com/jayway/jsonpath/spi/json/JsonProvider.java#L32) or [InputStream](https://github.com/jayway/JsonPath/blob/master/json-path/src/main/java/com/jayway/jsonpath/spi/json/JsonProvider.java#L41). 

I actually have  ProtobufProvider and ProtobufMapper implementation that takes DynamicMessage via InputStream.  But it's still not zero-copy and hacky.  Therefore I did not include them in this PR.

If we modify the JSONPath that takes `Object` type, we can implement zero copy provider.  And if we do, probably we have to maintain the fork.  I don't think JSONPath author would accept some random object provider that is not JSON ;-)